### ISSUE TITLE
ExecutorsPool should have an API to release Executor to pool

### ISSUE THREAD

**Issue by NikolayPianikov** [chechulnikov/eucalypt#11]

**Comment by chechulnikov** [comment]
Зачем?
Здесь такой дизайн, что пул для потребителя только отдаёт доступный резервируемый executor 
```kotlin
interface ExecutorsPool {
    suspend fun getAvailableExecutor(type: ExecutorType): Result<ReservableExecutor>
}
```
``` kotlin
interface ReservableExecutor : Executor {
    suspend fun tryReserve(): Boolean
    suspend fun release()
}
```
[`ExecutorManagerImpl`](https://github.com/chechulnikov/eucalypt/blob/main/src/src/main/kotlin/eucalypt/business/executing/ExecutorsManagerImpl.kt) уже пытается его зарезервировать и через него же можно "осободить" executor.
``` kotlin
interface ExecutorsManager {
    suspend fun borrowExecutor(type: ExecutorType): Result<Executor>
    suspend fun redeemExecutor(executor: Executor)
}
```

`ExecutorPool` для уровня исполнения скрипта недоступен, он работает с `ExecutorManager`.

Предполагалось, что выглядеть это будет так:
``` kotlin
...
executorsManager.borrowExecutor(type)
    .onSuccess { executor ->
        val executionResult = runOnExecutor(executor, script)
        executorsManager.redeemExecutor(executor)
        ...
    }
....
```

Если есть сомнения, то можно сделать одно из двух
- переименовать пул во что-то другое (не знаю пока во что), а `ExecutorManager` – в пул.
- внести логику ExecutorManager-а в пул (тогда и без того сложный пул станет ещё сложнее)

**Comment by NikolayPianikov** [comment]
Да, _ExecutorsManager_ имеет интерфес больше похожий на пулы чем сами пуды _ExecutorsPool_ :)
```
T GetObject();
void ReleaseObject(T obj);
```
И если бы не #15, то и задач у _ExecutorsManagerImpl_ не было бы.
Можно было бы _ExecutorsManager_ сделал пулом, его ответвенность была бы предоставлять объекты пула во временное использование, забирать их обратно и все!
- Сделал бы простую реализацию с коллекцией и базовой логикой
- Сделал бы реалзицию-декоратор, которая "прогревает" декорируемый пул исходя из некой конфигурации

Потребитель берет объекта из пула, временно владеет им и возвращает его в пул, когда он ему не нужен. Этот объект не занимается "самоанализом" и не возвращает себя в пул самостоятельно.

PS: я бы не делал пулы совсем, так как в данном случае они только усложняют логику, добавляют executors c состояниями, переходы между состояниями, синхронизацию между потоками, при этом не принося профита

**Comment by chechulnikov** [comment]
> PS: я бы не делал пулы совсем, так как в данном случае они только усложняют логику, добавляют executors c состояниями, переходы между состояниями, синхронизацию между потоками, при этом не принося профита, я бы сосредоточил свои усилия на другом ...

Если сделать реализацию без пула и с автоматическим ре-раном контейнеров, но тогда она бы состряпала из контроллера и класса, который раннит контейнер, передаёт скрипт, ждёт результат и возвращает ответ. Строчек 30.

Считаю такое задание бесполезным, т.к. не будет повода даже поговорить.
